src/pages/Environments.js

```diff
@@ -116,8 +120,165 @@ const DeleteAppModal = ({ app: { googleProject, appName, diskName, appType }, on
   ])
 }
 
+const MigratePersistentDisksBanner = ({ count }) => {
+  const deadline = new Date('01 January 2023 00:00 UTC')
+  return count > 0 && div({
+    style: {
+      position: 'absolute', top: topBarHeight, left: '50%', transform: 'translate(-50%, -50%)',
+      backgroundColor: colors.warning(0.15),
+      border: '2px solid', borderRadius: '12px', borderColor: colors.warning(),
+      zIndex: 8000 // Must be under 9000 to not draw over notifications
+    }
+  }, [
+    div({ style: { display: 'flex', alignItems: 'center', margin: '0.75rem 1.5rem 0.75rem 1.5rem' } }, [
+      icon('warning-standard', { size: 32, style: { color: colors.warning(), marginRight: '0.25rem' } }),
+      div([
+        strong([`You have ${Math.floor((deadline - Date.now()) / 86400000)} days to migrate ${count} shared persistent ${count > 1 ?
+          'disks' :
+          'disk'}. `]),
+        `Un-migrated disks will be DELETED after ${Utils.makeCompleteDate(deadline)}.`
+      ])
+    ])
+  ])
+}
+
+const MigratePersistentDiskCell = ({ disk, setMigrateDisk }) => div({
+  style: {
+    display: 'flex', flex: 1, flexDirection: 'column', textAlign: 'center',
+    height: '100%', margin: '-1rem', padding: '0.5rem 0 0 0',
+    backgroundColor: colors.danger(0.15), color: colors.danger()
+  }
+}, [
+  h(TooltipTrigger, { content: `This disk is shared between workspaces, which is no longer supported. Click "Migrate" to make copies for relevant workspaces.` }, [
+    div(['Offline', icon('warning-standard', { style: { marginLeft: '0.25rem', color: colors.danger() } })])
+  ]),
+  h(Link, { onClick: () => setMigrateDisk(disk), style: { wordBreak: 'break-word' } },
+    ['Migrate']
+  )
+])
+
+const MigratePersistentDiskModal = ({ disk, workspaces, onSuccess, onDismiss, contactSupport, deleteDiskId }) => {
+  const [migrating, setMigrating] = useState()
+  const [isSelected, setIsSelected] = useState({})
+  const [deleteDisk, setDeleteDisk] = useState(false)
+
+  const copyDiskToWorkspace =
+    ({ googleProject, namespace: saturnWorkspaceNamespace, name: saturnWorkspaceName }) => {
+      return reportErrorAndRethrow(`Error copying persistent disk to workspace '${saturnWorkspaceName}'`,
+        () => Ajax().Disks.disk(googleProject, disk.name).create({
+          ..._.pick(['size', 'blockSize', 'zone'], disk),
+          diskType: disk.diskType.label,
+          labels: { saturnWorkspaceNamespace, saturnWorkspaceName },
+          sourceDisk: _.pick(['googleProject', 'name'], disk)
+        })
+      )()
+    }
```

Comment:
Instead of immediately calling the function returned by `reportErrorAndRethrow`, it can be used like this:

```suggestion
  const copyDiskToWorkspace = reportErrorAndRethrow(
    `Error copying persistent disk to workspace '${saturnWorkspaceName}'`
    ({ googleProject, namespace: saturnWorkspaceNamespace, name: saturnWorkspaceName }) => {
      return Ajax().Disks.disk(googleProject, disk.name).create({
          ..._.pick(['size', 'blockSize', 'zone'], disk),
          diskType: disk.diskType.label,
          labels: { saturnWorkspaceNamespace, saturnWorkspaceName },
          sourceDisk: _.pick(['googleProject', 'name'], disk)
      })
    }
  )
```

Reply:
not sure how this works as the binding for `saturnWorkspaceName` comes from the ~lambda~ arrow?

Reply:
Oops, didn't notice that. Ignore this then.